# 01. Importing libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline

# 02. Importing data

In [2]:
# Assign the path of the Gun Violence data folder to the variable 'path'.
path = r'C:\Users\ramis\OneDrive\JC\Data Analytics Course\Data Analytics Immersion\A6\Gun Violence'

In [3]:
path

'C:\\Users\\ramis\\OneDrive\\JC\\Data Analytics Course\\Data Analytics Immersion\\A6\\Gun Violence'

In [4]:
# import gun_violence.csv
df = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'gun_violence.csv'))

In [5]:
df.head()

incident_id        date           state city_or_county  \
0       461105  2013-01-01    Pennsylvania     Mckeesport   
1       460726  2013-01-01      California      Hawthorne   
2       478855  2013-01-01            Ohio         Lorain   
3       478925  2013-01-05        Colorado         Aurora   
4       478959  2013-01-07  North Carolina     Greensboro   

                                     address  n_killed  n_injured  \
0  1506 Versailles Avenue and Coursin Street         0          4   
1               13500 block of Cerise Avenue         1          3   
2                      1776 East 28th Street         1          3   
3           16000 block of East Ithaca Place         4          0   
4                  307 Mourning Dove Terrace         2          2   

                                        incident_url  \
0  http://www.gunviolencearchive.org/incident/461105   
1  http://www.gunviolencearchive.org/incident/460726   
2  http://www.gunviolencearchive.org/incident/478855   
3  http://www.gunviolencearchive.org/incident/478925   
4  http://www.gunviolencearchive.org/incident/478959   

                                          source_url  \
0  http://www.post-gazette.com/local/south/2013/0...   
1  http://www.dailybulletin.com/article/zz/201301...   
2  http://chronicle.northcoastnow.com/2013/02/14/...   
3  http://www.dailydemocrat.com/20130106/aurora-s...   
4  http://www.journalnow.com/news/local/article_d...   

   incident_url_fields_missing  ...                    participant_age  \
0                        False  ...                              0::20   
1                        False  ...                              0::20   
2                        False  ...  0::25||1::31||2::33||3::34||4::33   
3                        False  ...         0::29||1::33||2::56||3::33   
4                        False  ...         0::18||1::46||2::14||3::47   

                               participant_age_group  \
0  0::Adult 18+||1::Adult 18+||2::Adult 18+||3::A...   
1  0::Adult 18+||1::Adult 18+||2::Adult 18+||3::A...   
2  0::Adult 18+||1::Adult 18+||2::Adult 18+||3::A...   
3  0::Adult 18+||1::Adult 18+||2::Adult 18+||3::A...   
4  0::Adult 18+||1::Adult 18+||2::Teen 12-17||3::...   

                            participant_gender  \
0         0::Male||1::Male||3::Male||4::Female   
1                                      0::Male   
2  0::Male||1::Male||2::Male||3::Male||4::Male   
3         0::Female||1::Male||2::Male||3::Male   
4       0::Female||1::Male||2::Male||3::Female   

                                    participant_name  \
0                                     0::Julian Sims   
1                                  0::Bernard Gillis   
2  0::Damien Bell||1::Desmen Noble||2::Herman Sea...   
3  0::Stacie Philbrook||1::Christopher Ratliffe||...   
4  0::Danielle Imani Jameison||1::Maurice Eugene ...   

   participant_relationship  \
0                       NaN   
1                       NaN   
2                       NaN   
3                       NaN   
4                 3::Family   

                                  participant_status  \
0  0::Arrested||1::Injured||2::Injured||3::Injure...   
1      0::Killed||1::Injured||2::Injured||3::Injured   
2  0::Injured, Unharmed, Arrested||1::Unharmed, A...   
3         0::Killed||1::Killed||2::Killed||3::Killed   
4       0::Injured||1::Injured||2::Killed||3::Killed   

                                    participant_type  \
0  0::Victim||1::Victim||2::Victim||3::Victim||4:...   
1  0::Victim||1::Victim||2::Victim||3::Victim||4:...   
2  0::Subject-Suspect||1::Subject-Suspect||2::Vic...   
3  0::Victim||1::Victim||2::Victim||3::Subject-Su...   
4  0::Victim||1::Victim||2::Victim||3::Subject-Su...   

                                             sources state_house_district  \
0  http://pittsburgh.cbslocal.com/2013/01/01/4-pe...                  NaN   
1  http://losangeles.cbslocal.com/2013/01/01/man-...                 62.0   
2  http://www.morningjournal.com/ge

# 03. Creating an alternative dataframe

In [6]:
# List of columns to drop
columns_to_drop = ['participant_relationship','location_description','participant_name','gun_stolen','n_guns_involved','gun_type','notes','state_house_district','state_senate_district','address','sources','source_url','incident_url_fields_missing','incident_url']

# Create a new dataframe df_gv by dropping the specified columns from df
df_gv = df.drop(columns=columns_to_drop, axis=1)

In [8]:
df_gv.head()

incident_id        date           state city_or_county  n_killed  \
0       461105  2013-01-01    Pennsylvania     Mckeesport         0   
1       460726  2013-01-01      California      Hawthorne         1   
2       478855  2013-01-01            Ohio         Lorain         1   
3       478925  2013-01-05        Colorado         Aurora         4   
4       478959  2013-01-07  North Carolina     Greensboro         2   

   n_injured  congressional_district  \
0          4                    14.0   
1          3                    43.0   
2          3                     9.0   
3          0                     6.0   
4          2                     6.0   

                            incident_characteristics  latitude  longitude  \
0  Shot - Wounded/Injured||Mass Shooting (4+ vict...   40.3467   -79.8559   
1  Shot - Wounded/Injured||Shot - Dead (murder, a...   33.9090  -118.3330   
2  Shot - Wounded/Injured||Shot - Dead (murder, a...   41.4455   -82.1377   
3  Shot - Dead (murder, accidental, suicide)||Off...   39.6518  -104.8020   
4  Shot - Wounded/Injured||Shot - Dead (murder, a...   36.1140   -79.9569   

                     participant_age  \
0                              0::20   
1                              0::20   
2  0::25||1::31||2::33||3::34||4::33   
3         0::29||1::33||2::56||3::33   
4         0::18||1::46||2::14||3::47   

                               participant_age_group  \
0  0::Adult 18+||1::Adult 18+||2::Adult 18+||3::A...   
1  0::Adult 18+||1::Adult 18+||2::Adult 18+||3::A...   
2  0::Adult 18+||1::Adult 18+||2::Adult 18+||3::A...   
3  0::Adult 18+||1::Adult 18+||2::Adult 18+||3::A...   
4  0::Adult 18+||1::Adult 18+||2::Teen 12-17||3::...   

                            participant_gender  \
0         0::Male||1::Male||3::Male||4::Female   
1                                      0::Male   
2  0::Male||1::Male||2::Male||3::Male||4::Male   
3         0::Female||1::Male||2::Male||3::Male   
4       0::Female||1::Male||2::Male||3::Female   

                                  participant_status  \
0  0::Arrested||1::Injured||2::Injured||3::Injure...   
1      0::Killed||1::Injured||2::Injured||3::Injured   
2  0::Injured, Unharmed, Arrested||1::Unharmed, A...   
3         0::Killed||1::Killed||2::Killed||3::Killed   
4       0::Injured||1::Injured||2::Killed||3::Killed   

                                    participant_type  
0  0::Victim||1::Victim||2::Victim||3::Victim||4:...  
1  0::Victim||1::Victim||2::Victim||3::Victim||4:...  
2  0::Subject-Suspect||1::Subject-Suspect||2::Vic...  
3  0::Victim||1::Victim||2::Victim||3::Subject-Su...  
4  0::Victim||1::Victim||2::Victim||3::Subject-Su...

# 04 Data cleaning

In [11]:
# details of he dataframe
df_gv.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 239677 entries, 0 to 239676
Data columns (total 15 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   incident_id               239677 non-null  int64  
 1   date                      239677 non-null  object 
 2   state                     239677 non-null  object 
 3   city_or_county            239677 non-null  object 
 4   n_killed                  239677 non-null  int64  
 5   n_injured                 239677 non-null  int64  
 6   congressional_district    227733 non-null  float64
 7   incident_characteristics  239351 non-null  object 
 8   latitude                  231754 non-null  float64
 9   longitude                 231754 non-null  float64
 10  participant_age           147379 non-null  object 
 11  participant_age_group     197558 non-null  object 
 12  participant_gender        203315 non-null  object 
 13  participant_status        212051 non-null  o

In [12]:
# shape of the dataframe
df_gv.shape

(239677, 15)

In [13]:
# descriptive statistics of the dataframe
df_gv.describe()

incident_id       n_killed      n_injured  congressional_district  \
count  2.396770e+05  239677.000000  239677.000000           227733.000000   
mean   5.593343e+05       0.252290       0.494007                8.001265   
std    2.931287e+05       0.521779       0.729952                8.480835   
min    9.211400e+04       0.000000       0.000000                0.000000   
25%    3.085450e+05       0.000000       0.000000                2.000000   
50%    5.435870e+05       0.000000       0.000000                5.000000   
75%    8.172280e+05       0.000000       1.000000               10.000000   
max    1.083472e+06      50.000000      53.000000               53.000000   

            latitude      longitude  
count  231754.000000  231754.000000  
mean       37.546598     -89.338348  
std         5.130763      14.359546  
min        19.111400    -171.429000  
25%        33.903400     -94.158725  
50%        38.570600     -86.249600  
75%        41.437375     -80.048625  
max        71.336800      97.433100

In [9]:
# Check for missing values in each column of df_gv
missing_values = df_gv.isnull().sum()
# Display only columns with missing values
missing_values[missing_values > 0]

congressional_district      11944
incident_characteristics      326
latitude                     7923
longitude                    7923
participant_age             92298
participant_age_group       42119
participant_gender          36362
participant_status          27626
participant_type            24863
dtype: int64

In [15]:
# Calculate the proportion of missing values for participant_age_group
total_rows = df_gv.shape[0]
participant_age_group_missing_proportion = missing_values['participant_age_group'] / total_rows*100
participant_age_group_missing_proportion

17.573233977394576

####  As we can see above, the proportion of missing values for participant_age_group is approx. 17.57%. The values in this column appear to be coded in a specific format, where each entry may consist of one or more age group designations, each prefixed with an identifier (like 0::, 1::, etc.), which may correspond to different participants. Thus, it would be very tricky to try to imputate the missing values. My approach would consider grouping the data into 'Known Age Group' and 'Unknown Age Group' to retain the records for certain types of analysis.

In [16]:
# Group the data into 'Known Age Group' and 'Unknown Age Group' in a new column called 'age_group_classification'
df_gv['age_group_classification'] = df_gv['participant_age_group'].apply(lambda x: 'Known Age Group' if pd.notnull(x) else 'Unknown Age Group')

# Display the counts for each classification to verify that we categorized the records correctly
age_group_classification_counts = df_gv['age_group_classification'].value_counts()
age_group_classification_counts

age_group_classification
Known Age Group      197558
Unknown Age Group     42119
Name: count, dtype: int64

In [20]:
# check how the new column 'age_group_classification' and its values look like
df_gv.head()

incident_id        date           state city_or_county  n_killed  \
0       461105  2013-01-01    Pennsylvania     Mckeesport         0   
1       460726  2013-01-01      California      Hawthorne         1   
2       478855  2013-01-01            Ohio         Lorain         1   
3       478925  2013-01-05        Colorado         Aurora         4   
4       478959  2013-01-07  North Carolina     Greensboro         2   

   n_injured  congressional_district  \
0          4                    14.0   
1          3                    43.0   
2          3                     9.0   
3          0                     6.0   
4          2                     6.0   

                            incident_characteristics  latitude  longitude  \
0  Shot - Wounded/Injured||Mass Shooting (4+ vict...   40.3467   -79.8559   
1  Shot - Wounded/Injured||Shot - Dead (murder, a...   33.9090  -118.3330   
2  Shot - Wounded/Injured||Shot - Dead (murder, a...   41.4455   -82.1377   
3  Shot - Dead (murder, accidental, suicide)||Off...   39.6518  -104.8020   
4  Shot - Wounded/Injured||Shot - Dead (murder, a...   36.1140   -79.9569   

                     participant_age  \
0                              0::20   
1                              0::20   
2  0::25||1::31||2::33||3::34||4::33   
3         0::29||1::33||2::56||3::33   
4         0::18||1::46||2::14||3::47   

                               participant_age_group  \
0  0::Adult 18+||1::Adult 18+||2::Adult 18+||3::A...   
1  0::Adult 18+||1::Adult 18+||2::Adult 18+||3::A...   
2  0::Adult 18+||1::Adult 18+||2::Adult 18+||3::A...   
3  0::Adult 18+||1::Adult 18+||2::Adult 18+||3::A...   
4  0::Adult 18+||1::Adult 18+||2::Teen 12-17||3::...   

                            participant_gender  \
0         0::Male||1::Male||3::Male||4::Female   
1                                      0::Male   
2  0::Male||1::Male||2::Male||3::Male||4::Male   
3         0::Female||1::Male||2::Male||3::Male   
4       0::Female||1::Male||2::Male||3::Female   

                                  participant_status  \
0  0::Arrested||1::Injured||2::Injured||3::Injure...   
1      0::Killed||1::Injured||2::Injured||3::Injured   
2  0::Injured, Unharmed, Arrested||1::Unharmed, A...   
3         0::Killed||1::Killed||2::Killed||3::Killed   
4       0::Injured||1::Injured||2::Killed||3::Killed   

                                    participant_type age_group_classification  
0  0::Victim||1::Victim||2::Victim||3::Victim||4:...          Known Age Group  
1  0::Victim||1::Victim||2::Victim||3::Victim||4:...          Known Age Group  
2  0::Subject-Suspect||1::Subject-Suspect||2::Vic...          Known Age Group  
3  0::Victim||1::Victim||2::Victim||3::Subject-Su...          Known Age Group  
4  0::Victim||1::Victim||2::Victim||3::Subject-Su...          Known Age Group

In [21]:
# export df_gv to csv to save the changes we've done so far
df_gv.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'gun_violence_c1.csv'))

In [25]:
# Check the data types of each column to ensure they are appropriate for the data they contain
df_gv.dtypes

incident_id                   int64
date                         object
state                        object
city_or_county               object
n_killed                      int64
n_injured                     int64
congressional_district      float64
incident_characteristics     object
latitude                    float64
longitude                   float64
participant_age              object
participant_age_group        object
participant_gender           object
participant_status           object
participant_type             object
age_group_classification     object
dtype: object

#### Based on the data types shown above, the data type in the 'date' column is currently object. This needs to be corrected to datetime64. I will perform this change below

In [26]:
# convert data type of the 'date' column to datetime
df_gv['date'] = pd.to_datetime(df_gv['date'])

In [28]:
# check if data type is now correct for the 'date' column
df_gv.dtypes

incident_id                          int64
date                        datetime64[ns]
state                               object
city_or_county                      object
n_killed                             int64
n_injured                            int64
congressional_district             float64
incident_characteristics            object
latitude                           float64
longitude                          float64
participant_age                     object
participant_age_group               object
participant_gender                  object
participant_status                  object
participant_type                    object
age_group_classification            object
dtype: object

In [29]:
# check for duplicate rows
duplicate_rows = df_gv.duplicated().sum()
duplicate_rows

0

#### based on the above, we have no duplicates.

In [32]:
# Examine the following columns  for unique values to spot any anomalies or unusual entries (e.g., a state name that is misspelled)
# Specify the columns I want to check
columns_to_check = ['state', 'n_killed', 'n_injured']

# Get unique values for each of the specified columns
unique_values = {col: df_gv[col].unique() for col in columns_to_check}
unique_values

{'state': array(['Pennsylvania', 'California', 'Ohio', 'Colorado', 'North Carolina',
        'Oklahoma', 'New Mexico', 'Louisiana', 'Maryland', 'Tennessee',
        'Missouri', 'District of Columbia', 'Illinois', 'Delaware', 'Utah',
        'Michigan', 'Georgia', 'Indiana', 'Mississippi', 'New York',
        'Florida', 'Washington', 'South Carolina', 'Arizona', 'Kentucky',
        'New Jersey', 'Virginia', 'Wisconsin', 'Rhode Island', 'Texas',
        'Alabama', 'Kansas', 'Connecticut', 'West Virginia', 'Minnesota',
        'Nevada', 'Nebraska', 'Massachusetts', 'Hawaii', 'New Hampshire',
        'Iowa', 'Alaska', 'Arkansas', 'Idaho', 'Oregon', 'Wyoming',
        'Maine', 'North Dakota', 'Montana', 'Vermont', 'South Dakota'],
       dtype=object),
 'n_killed': array([ 0,  1,  4,  2,  5,  3,  6,  7, 11,  8,  9, 10, 16, 50, 27, 17],
       dtype=int64),
 'n_injured': array([ 4,  3,  0,  2,  5,  6,  1,  8, 13,  7, 19,  9, 12, 16, 15, 18, 11,
        10, 17, 14, 53, 25, 20], dtype=int64)}

#### no issues noticed in the outcome above.

# 05 Exporting the clean dataframe to pkl

In [33]:
# export df_gv to pkl to save the changes we've done so far and to preserve the data types of the columns
df_gv.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'gun_violence_c2.pkl'))